config/application.rb

```diff
@@ -43,7 +43,7 @@ class Application < Rails::Application
           ServiceProvider.pluck(:redirect_uris).flatten.compact.map do |uri|
             begin
               URI.join(uri, '/').to_s[0..-2]
-            rescue URI::BadURIError => err
+            rescue URI::Error => err
               Rails.logger.warn({ warning: err.class.to_s, source: 'Rack::Cors', uri: uri }.to_json)
               ''
```

Comment:
Should we also, or instead of Rails.logger, send this to the exception notification?
```ruby
ExceptionNotifier.notify_exception(err)
```

Reply:
My only concern would be about noise. Currently we have a lot of invalid URIs so we're expecting this to raise a ton of errors on every post request.

Reply:
I think this is a change that could fit in #1762 though

Reply:
Good point about the noise, which made me look deeper into this and wonder why we are running this code on every POST request when it only applies to CORS requests. What if we add this at the top of the `origins` block?

```ruby
next if source.include?(Figaro.env.domain_name)
```
That way, we don't call the DB on every request, and then simplify the logic to this:
```ruby
ServiceProvider.pluck(:redirect_uris).flatten.compact.find do |uri|
  uri.sub(/(\/)+$/,'').include?(source)
end
```
I'm the one who originally added the `URI.join` code, and I don't remember exactly why, but I suspect it's because I was trying to deal with the scenario where the redirect_uri in the DB had a forward slash at the end, and because I was doing `Array.include?(source)`, which won't find a match if the source doesn't have the trailing slash. That was the wrong comparison to use. 

Can you think of another reason why we would want to use URI at all if all we want to do is make sure the HTTP_ORIGIN matches one of the `redirect_uri`s in the DB (minus any trailing slashes)?

I ran a benchmark and the regex is 22x faster than URI parsing:

```ruby
require 'benchmark/ips'
require 'uri'

URIS = [
  'gov.dhs.cbp.pspd.oars.user.prod://result',
  'http://test.com/',
  'https://careers.cbp.dhs.gov/hrm/app'
].freeze


def with_regex
  URIS.find { |uri| uri.sub(/(\/)+$/,'').include?('http://test.com') }
end

def with_uri
  URIS.map do |uri|
    begin
      URI.join(uri, '/').to_s[0..-2]
    rescue URI::BadURIError => err
      ''
    end
  end.include?('http://test.com')
end

Benchmark.ips do |x|
  x.report("regex")  { with_regex }
  x.report("uri") { with_uri }
  x.compare!
end
```

```
Warming up --------------------------------------
               regex    19.931k i/100ms
                 uri     1.011k i/100ms
Calculating -------------------------------------
               regex    241.670k (± 3.5%) i/s -      1.216M in   5.036934s
                 uri     10.604k (± 4.3%) i/s -     53.583k in   5.063209s

Comparison:
               regex:   241670.5 i/s
                 uri:    10603.8 i/s - 22.79x slower
```

Reply:
Using URI.join let's us remove the path from the redirect URI. That way we can use exact matching to check the source.

```ruby
uri = 'https://careers-cert.cbp.dhs.gov/hrm/app'
URI.join(uri, '/').to_s[0..-2]
# => https://careers-cert.cbp.dhs.gov
```

I'm not sure the fuzzy matching is what we want since it is easy to makes mistakes. For example, if we have a service provider with the redirect uri `https://asdf.dod.gov`, someone could register `http://asdf.do` and our app would recognize that as a trusted origin.

Reply:
Also, I think we can perhaps avoid going to the database for all the service providers on every request here, but also that adds some complexity to the bug fix in this PR. I'd prefer to come back and address this in a follow-on PR.

Reply:
How about using `split`, like we do in https://github.com/18F/identity-idp/blob/master/app/services/secure_headers_whitelister.rb ?
```ruby
ServiceProvider.pluck(:redirect_uris).flatten.compact.find do |uri|
  split_uri = uri.split('//')
  protocol = split_uri[0]
  domain = split_uri[1].split('/')[0]
  "#{protocol}//#{domain}" == source
end
```

Reply:
That might work once we start validating them, but won't now since we have values that are empty strings. If we can, I'd like to focus on fixing the bug in this PR, and then move discussion about enhancements to a follow-on (#1762) where we can spend more time discussing it.

Reply:
@monfresh: I went ahead and added some of the changes we talked about here to #1762. PTAL

Reply:
Sounds good.